src/server/transaction/server/driver.go

```diff
@@ -0,0 +1,292 @@
+				response = client.NewCommitResponse(commit)
+			}
+		case *transaction.TransactionRequest_FinishCommit:
+			err = d.txnEnv.PfsServer().FinishCommitInTransaction(pachClient, stm, x.FinishCommit)
+			response = client.NewEmptyResponse()
+		case *transaction.TransactionRequest_DeleteCommit:
+			err = d.txnEnv.PfsServer().DeleteCommitInTransaction(pachClient, stm, x.DeleteCommit)
+			response = client.NewEmptyResponse()
+		case *transaction.TransactionRequest_CreateBranch:
+			err = d.txnEnv.PfsServer().CreateBranchInTransaction(pachClient, stm, x.CreateBranch)
+			response = client.NewEmptyResponse()
+		case *transaction.TransactionRequest_DeleteBranch:
+			err = d.txnEnv.PfsServer().DeleteBranchInTransaction(pachClient, stm, x.DeleteBranch)
+			response = client.NewEmptyResponse()
+		case *transaction.TransactionRequest_CopyFile:
+			err = d.txnEnv.PfsServer().CopyFileInTransaction(pachClient, stm, x.CopyFile)
+			response = client.NewEmptyResponse()
+		case *transaction.TransactionRequest_DeleteFile:
+			err = d.txnEnv.PfsServer().DeleteFileInTransaction(pachClient, stm, x.DeleteFile)
+			response = client.NewEmptyResponse()
+		case *transaction.TransactionRequest_DeleteAll:
+			// TODO: extend this to delete everything through PFS, PPS, Auth and
+			// update the client DeleteAll call to use only this, then remove unused
+			// RPCs.
+			err = d.deleteAll(pachClient, stm, info.Transaction)
+			response = client.NewEmptyResponse()
+		default:
+			err = fmt.Errorf("unrecognized transaction request type")
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("error running request %d of %d: %v", i+1, len(info.Requests), err)
+		}
+
+		responses = append(responses, response)
+	}
+
+	info.Responses = responses
+	return info, nil
+}
+
+func (d *driver) finishTransaction(pachClient *client.APIClient, txn *transaction.Transaction) (*transaction.TransactionInfo, error) {
+	ctx := pachClient.Ctx()
+	info := &transaction.TransactionInfo{}
+	_, err := col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+		err := d.transactions.ReadOnly(ctx).Get(txn.ID, info)
+		if err != nil {
+			return err
+		}
+		info, err = d.runTransaction(pachClient, stm, info)
+		if err != nil {
+			return err
+		}
+		return d.transactions.ReadWrite(stm).Delete(txn.ID)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return info, nil
+}
+
+// Error to be returned when the transaction has been modified between our two STM calls
+type transactionConflictError struct{}
+
+func (e *transactionConflictError) Error() string {
+	return fmt.Sprintf("transaction could not be modified due to concurrent modifications")
+}
+
+func (d *driver) appendTransaction(
+	pachClient *client.APIClient,
+	txn *transaction.Transaction,
+	items []*transaction.TransactionRequest,
+) (*transaction.TransactionInfo, error) {
+	ctx := pachClient.Ctx()
+
+	// Run this thing in a loop in case we get a conflict, time out after some tries
+	for i := 0; i < 10; i++ {
```

Comment:
It might be good the have the number of retries be configurable in the request. Just in case there's a situation where people have a lot of traffic and need this to be higher.

Reply:
Unfortunately because of [this change](https://github.com/pachyderm/pachyderm/pull/3658#discussion_r278336605), we won't be able to specify the number of retries on each request, it will need to be specified when creating or finishing the transaction.  I could add another `update transaction` command that could set this value, but it seems like overkill - thoughts?

Reply:
Yeah that seems like overkill, and configurable retries doesn't actually seem that important thinking about it more.